lib/pharos/up_command.rb

```diff
@@ -77,27 +77,39 @@ def configure(config)
       master_hosts = master_hosts(config)
       signal_usage_error 'No master hosts defined' if master_hosts.empty?
       signal_usage_error 'Only one host can be in master role' if master_hosts.size > 1
-
-      begin
-        start_time = Time.now
-        puts pastel.green("==> Sharpening tools ...")
-        load_phases
-        puts pastel.green("==> Starting to craft cluster ...")
-        validate_hosts(config)
-        # set workdir to the same dir where config was loaded from
-        # so that the certs etc. can be referenced more easily
-        Dir.chdir(config_yaml.dirname) do
-          handle_masters(master_hosts[0], config)
-          handle_workers(master_hosts[0], worker_hosts(config), config)
-          handle_addons(master_hosts[0], config.addons)
-        end
-        craft_time = Time.now - start_time
-        puts pastel.green("==> Cluster has been crafted! (took #{humanize_duration(craft_time.to_i)})")
-        puts "    You can connect to the cluster with kubectl using:"
-        puts "    export KUBECONFIG=~/.pharos/#{master_hosts[0].address}"
-      ensure
-        Pharos::SSH::Client.disconnect_all
+      master_host = master_hosts[0]
+
+      start_time = Time.now
+      puts pastel.green("==> Sharpening tools ...")
+      load_phases
+      puts pastel.green("==> Starting to craft cluster ...")
+
+      # set workdir to the same dir where config was loaded from
+      # so that the certs etc. can be referenced more easily
+      Dir.chdir(config_yaml.dirname) do
+        handle_phase(Phases::ValidateHost, config.hosts, config: config)
+        handle_phase(Phases::ConfigureHost, config.hosts, config: config)
+
+        handle_phase(Phases::ConfigureMaster, master_hosts, config: config, master: master_host)
```

Comment:
these could be something like:
`queue << Phases::ConfigureHost.new(config.hosts, config: config)`

and the thing that runs the phases / queue could then just expect that it can `queue.pop.call` them without any args.

this way it would be possible also to do something like:

```
queue.later do 
  do stuff
end
```

and it would magically work.

Also then you wouldn't have to normalize the phases, you can just `queue << FooPhase.new(whatever_args_it_wants)`

Reply:
Not sure how putting them into a `Queue` would be any better than just calling them normally from a function? All the phases run in serial, it's much simpler that way... there's no need to parallelize the phases.

Reply:
queue was not capitalized.

but yes, no need for a queue there, but the thing that does the `.call` wouldn't then have to do the `.new` and play with the arguments.

Reply:
```
handle_phase(Phases:Foofoo.new(xyz))
```

Reply:
Yeah, that part is mostly about abstracting the `ssh` away, without needing to defer it until after `new`.

However, then there's also the question of what the abstraction should be for e.g. phases that  don't need `SSH`, or even those that are not node-specific, but apply to the entire kube cluster (e.g. `ConfigureNetwork`). However, that is also going to need bigger changes for multi-master support, and I don't know what those are going to look like yet... the `Pharos::Kube.apply_stack(@master.address, ...)` has to change as well.